pkg/ocmagenthandler/ocmagenthandler_deployment.go

```diff
@@ -0,0 +1,266 @@
+						Command:      ocmAgentCommand,
+						Name:         oahconst.OCMAgentName,
+						Ports: []corev1.ContainerPort{{
+							ContainerPort: oahconst.OCMAgentPort,
+							Name:          oahconst.OCMAgentPortName,
+						}},
+					}},
+				},
+			},
+		},
+	}
+	return dep
+}
+
+// buildOCMAgentArgs returns the full command argument list to run the OCM Agent
+// in a deployment.
+func buildOCMAgentArgs(ocmAgent ocmagentv1alpha1.OcmAgent) []string {
+	accessTokenPath := filepath.Join(oahconst.OCMAgentSecretMountPath, ocmAgent.Spec.TokenSecret,
+		oahconst.OCMAgentAccessTokenSecretKey)
+	configServicesPath := filepath.Join(oahconst.OCMAgentConfigMountPath, ocmAgent.Spec.OcmAgentConfig,
+		oahconst.OCMAgentConfigServicesKey)
+	configURLPath := filepath.Join(oahconst.OCMAgentConfigMountPath, ocmAgent.Spec.OcmAgentConfig,
+		oahconst.OCMAgentConfigURLKey)
+
+	command := []string {
+		oahconst.OCMAgentCommand,
+		"serve",
+		fmt.Sprintf("--access-token=@%s", accessTokenPath),
+		fmt.Sprintf("--services=@%s", configServicesPath),
+		fmt.Sprintf("--ocm-url=@%s", configURLPath),
+	}
+	return command
+}
+
+// ensureDeployment ensures that an OCMAgent Deployment exists on the cluster
+// and that its configuration matches what is expected.
+func (o *ocmAgentHandler) ensureDeployment(ocmAgent ocmagentv1alpha1.OcmAgent) error {
+	namespacedName := oahconst.BuildNamespacedName()
+	foundResource := &appsv1.Deployment{}
+	populationFunc := func() appsv1.Deployment {
+		return buildOCMAgentDeployment(ocmAgent)
+	}
+	// Does the resource already exist?
+	if err := o.Client.Get(o.Ctx, namespacedName, foundResource); err != nil {
+		if k8serrors.IsNotFound(err) {
+			// It does not exist, so must be created.
+			o.Log.Info("An OCMAgent deployment does not exist; will be created.")
+			// Populate the resource with the template
+			resource := populationFunc()
+			// Set the controller reference
+			if err := controllerutil.SetControllerReference(&ocmAgent, &resource, o.Scheme); err != nil {
+				return err
+			}
+			// and create it
+			err = o.Client.Create(o.Ctx, &resource)
+			if err != nil {
+				return err
+			}
+		} else {
+			// Return unexpectedly
+			return err
+		}
+	} else {
+		// It does exist, check if it is what we expected
+		resource := populationFunc()
+		if deploymentConfigChanged(foundResource, &resource, o.Log) {
+			// Specs aren't equal, update and fix.
+			o.Log.Info("An OCMAgent deployment exists but contains unexpected configuration. Restoring.")
+			foundResource.Spec = *resource.Spec.DeepCopy()
+			if err = o.Client.Update(o.Ctx, foundResource); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// ensureDeploymentDeleted removes the deployment from the cluster
+func (o *ocmAgentHandler) ensureDeploymentDeleted() error {
+	namespacedName := oahconst.BuildNamespacedName()
+	foundResource := &appsv1.Deployment{}
+	// Does the resource already exist?
+	if err := o.Client.Get(o.Ctx, namespacedName, foundResource); err != nil {
```

Comment:
[future work maybe] Can we have a pkg/ocmagenthandler/helper.go here? So that we can abstract all the steps like this with a simple look.

Reply:
I had an attempt at this yesterday in a separate branch:
https://github.com/mrbarge/ocm-agent-operator/commit/de059dc41061fc439fad90b02119a74b398ae070

It works, but going back and forth between `Unstructured` type to accomplish this really doesn't sit very well with me, it makes part of the code harder to understand. I think it is maybe something to consider for future work though.